include/tvm/target/compilation_config.h

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/target/compilation_config.h
+ * \brief A helper class to collect all the targets in canonical form necessary for compilation.
```

Comment:
Do not need to address in this PR given this is still experimental, but would be useful to have a followup discussion on:

- A0: rely on most configs in PassContext and IRModule attachment(target constraint of each functions).
- A1: centralize options in a single structure.

We will need to think about strategies in A0 and A1 and how do they interact with each other. 

If we are building a fixed function, closed box toolkit, then A1 is usually sufficient. 

In our case, to enable open box philosophy, we need to consider cases where constraints are pre-populated by passes not written by us(e.g. BYOC to CUDA that only works for cuda), and iterative refinement over the process.  In that case, we want IRModule to be self-sufficient for constraints that are already populated, and make followup build function respect them.

Reply:
See https://discuss.tvm.apache.org/t/pre-rfc-compilation-configuration-representation/11372